### PR TITLE
Mannually-augmented and weakly-supervised mrc model for DROP dataset

### PR DESCRIPTION
**Pull request by yizhongw**

### CODE REVIEW

File: reading_comprehension/qanet_for_drop.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+        self._span_accuracy = BooleanAccuracy()
+        self._squad_metrics = SquadEmAndF1()
+        self._dropout = torch.nn.Dropout(p=dropout_prob)
+
+        initializer(self)
+
+    def forward(self,  # type: ignore
+                question: Dict[str, torch.LongTensor],
+                passage: Dict[str, torch.LongTensor],
+                numbers_in_passage: Dict[str, torch.LongTensor],
```
Comment (matt-gardner) [inline reading_comprehension/qanet_for_drop.py:91]: I'd recommend having this be an `IndexField` instead of a `TextField`.  This way you just _point_ to the positions of the numbers, and it should be easier on the model.  I'll comment more on this below.

File: reading_comprehension/qanet_for_drop.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        # Shape: (batch_size, passage_length)
+        span_end_probs = masked_softmax(span_end_logits, passage_mask)
+
+        # Shape: (batch_size, # of numbers in the passage)
+        numbers_mask = util.get_text_field_mask(numbers_in_passage).float()
+        # Shape: (batch_size, # of numbers in the passage, embedding_dim)
+        embedded_numbers = self._dropout(self._text_field_embedder(numbers_in_passage))
+        embedded_numbers = self._number_embedding_proj_layer(embedded_numbers)
+        encoded_numbers = \
+            torch.cat([embedded_numbers, passage_vector.unsqueeze(1).repeat(1, embedded_numbers.size(1), 1)], -1)
```
Comment (matt-gardner) [inline reading_comprehension/qanet_for_drop.py:171]: Instead of this concatenation, I was imagining something that does a linear prediction on the encoded representation of the passage index.  So, e.g.:

```python
encoded_numbers = modeled_passage_list[-1][number_indices]  # indexing isn't quite right here, but hopefully the idea makes sense
number_sign_logits = self._number_sign_predictor(encoded_numbers)
```

Does this make sense?  Seems easier for the model to just encode somewhere for each number whether it belongs in the computation, instead of having to remember that through a projection to a single passage vector.

File: reading_comprehension/qanet_for_drop.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+        passage_weights = self._passage_weights_predictor(modeled_passage_list[4]).squeeze(-1)
+        passage_weights = passage_weights * passage_mask
+        passage_vector = util.weighted_sum(modeled_passage_list[4], passage_weights)
+        question_weights = self._question_weights_predictor(encoded_question).squeeze(-1)
+        question_weights = question_weights * question_mask
+        question_vector = util.weighted_sum(encoded_question, question_weights)
+
+        # Shape: (batch_size, 3)
+        answer_type_logits = self._answer_type_predictor(torch.cat([passage_vector, question_vector], -1))
+        answer_type_probs = torch.nn.functional.softmax(answer_type_logits, -1)
```
Comment (matt-gardner) [inline reading_comprehension/qanet_for_drop.py:150]: Is it possible to track accuracy on this?  That'd be helpful.

File: reading_comprehension/qanet_for_drop.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+            # Shape: (batch_size, # of numbers in the passage, # of combinations)
+            combinations = answer_as_plus_minus_combinations.transpose(1, 2)
+            # Shape: (batch_size, # of numbers in the passage, # of combinations)
+            likelyhood_for_combination_numbers = torch.gather(number_sign_probs, 2, combinations)
+            # the likelyhood of the masked positions should be 1 so that it will not affect the joint probability
+            likelyhood_for_combination_numbers = \
+                likelyhood_for_combination_numbers * numbers_mask.unsqueeze(-1) + (1 - numbers_mask.unsqueeze(-1))
+
+            # Shape: (batch_size, # of combinations)
+            likelyhood_for_combinations = \
```
Comment (matt-gardner) [inline reading_comprehension/qanet_for_drop.py:244]: This is not numerically stable.  If you really want to work in probability space, use `util.logsumexp()`.  But really, all of this logic should be done in log space, so you don't have underflow problems.

File: reading_comprehension/qanet_for_drop.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+                (likelyhood_for_combination_numbers + 1e-10).log().sum(1).exp() * combination_mask
+
+            # Some count answers are padded with label -1, we should mask them
+            # Shape: (batch_size, # of count answers)
+            count_mask = (answer_as_counts != -1).float()
+            # Shape: (batch_size, # of count answers)
+            clamped_counts = torch.nn.functional.relu(answer_as_counts)
+            likelyhood_for_counts = torch.gather(count_number_probs, 1, clamped_counts) * count_mask
+            # Shape: (batch_size, )
+            marginal_likelyhood = answer_type_probs[:, 0] * likelyhood_for_spans.sum(-1) \
```
Comment (matt-gardner) [inline reading_comprehension/qanet_for_drop.py:254]: It should be "likelihood", not "likelyhood", throughout.

**Review by matt-gardner — COMMENTED** [review]
The loss computation looks generally correct, but I'm guessing you're getting underflow problems because you're doing it all in probability space instead of log probability space.  I'd just work with log probabilities everywhere, and use `nn.util.logsumexp()` where appropriate.

File: reading_comprehension/qanet_for_drop.py (hunk 2 above)
Comment (yizhongw) [reply reading_comprehension/qanet_for_drop.py:171]: Yeah, this makes a lot of sense. I will do this.

File: reading_comprehension/qanet_for_drop.py (hunk 3 above)
Comment (yizhongw) [reply reading_comprehension/qanet_for_drop.py:150]: Ok, I will write some codes to track this.  For analysis, I have printed some results for this, it seems that the model predict all types as `span answer`.

File: reading_comprehension/qanet_for_drop.py (hunk 4 above)
Comment (yizhongw) [reply reading_comprehension/qanet_for_drop.py:244]: I'm not sure about this suggestion. `util.logsumexp()` equals to `tensor.exp().sum().log()`, while I want to do `tensor.log().sum().exp()`. This is because I need to multiply those likelihoods, and it's easier for pytorch to do it if I transform multiply to addition using `log`. I cannot keep those likelihoods in log space, since I need to add those likelihoods when computing the marginal likelihood (lines 254~257). Does this make sense? Any further suggestions?

Comment (matt-gardner) [reply reading_comprehension/qanet_for_drop.py:244]: Oh, sorry, you're right about this line.  But you should definitely just keep everything in log space.

You should use `util.logsumexp()` when you need to _add_ log numbers in probability space.  So, lines 254-257 should use `util.logsumexp()` to do the sum.

Comment (matt-gardner) [reply reading_comprehension/qanet_for_drop.py:244]: Basically, when you go from probability space to log space, you replace `*` everywhere with `+`, and `+` with `logsumexp()`.

Comment (yizhongw) [reply reading_comprehension/qanet_for_drop.py:244]: But what I need to do in lines 254-257 is to add probabilities rather than log probabilities, right? The loss is a negtive marginal likelihood, like:
```
loss = - log( P(type=span) * \sum P(span_i) + P(type=number) * \sum P(signs_for_numbers) + P(type=count) * \sum P(count) )
```

Comment (matt-gardner) [reply reading_comprehension/qanet_for_drop.py:244]: Yes, and the way that you add probabilities, when the probabilities are given in log space, is to use `logsumexp()`:

```python
# all of the log_likelihood_for_* variables contain _log probabilities_
marginal_span_log_likelihood = answer_type_probs[:, 0] + util.logsumexp(log_likelihood_for_spans, dim=-1)
marginal_combination_log_likelihood = answer_type_probs[: 1] + util.logsumexp(log_likelihood_for_combinations, dim=-1)
marginal_count_log_likelihood = answer_type_probs[:, 2] + util.logsumexp(log_likelihood_for_counts, dim=-1)
marginal_log_likelihood = util.logsumexp(torch.stack([marginal_span_log_likelihood, marginal_combination_log_likelihood, marginal_count_log_likelihood], dim=-1), dim=-1)
```

At least that's the basic idea; might be some bugs in there.

Comment (yizhongw) [reply reading_comprehension/qanet_for_drop.py:244]: Oh, I see the point here. Thx!
